impl/sql/adminstorage/admin_storage_test.go

```diff
@@ -0,0 +1,199 @@
+					VRFPriv: &keyspb.PrivateKey{Der: []byte("privkeybytes")},
+				},
+				{
+					Domain:  "domain2",
+					MapID:   1,
+					LogID:   2,
+					VRF:     &keyspb.PublicKey{Der: []byte("pubkeybytes")},
+					VRFPriv: &keyspb.PrivateKey{Der: []byte("privkeybytes")},
+				},
+			},
+		},
+	} {
+		for _, d := range tc.domains {
+			if err := admin.Write(ctx, d.Domain, d.MapID, d.LogID, d.VRF.Der, d.VRFPriv); err != nil {
+				t.Errorf("Write(): %v", err)
+				continue
+			}
+		}
+
+		domains, err := admin.List(ctx, tc.readDeleted)
+		if err != nil {
+			t.Errorf("List(): %v", err)
+			continue
+		}
+		if got, want := domains, tc.domains; !reflect.DeepEqual(got, want) {
+			t.Errorf("Domain: %v, want %v", got, want)
+		}
+	}
+}
+
+func TestWriteReadDelete(t *testing.T) {
+	ctx := context.Background()
+	db, err := sql.Open("sqlite3", ":memory:")
+	if err != nil {
+		t.Fatalf("sql.Open(): %v", err)
+	}
+	defer db.Close()
+	admin, err := New(db)
+	if err != nil {
+		t.Fatalf("Failed to create adminstorage: %v", err)
+	}
+
+	for _, tc := range []struct {
+		desc                 string
+		d                    adminstorage.Domain
+		write                bool
+		wantWriteErr         bool
+		setDelete, isDeleted bool
+		readDeleted          bool
+		wantReadErr          bool
+	}{
+		{
+			desc:  "Success",
+			write: true,
+			d: adminstorage.Domain{
+				Domain:  "testdomain",
+				MapID:   1,
+				LogID:   2,
+				VRF:     &keyspb.PublicKey{Der: []byte("pubkeybytes")},
+				VRFPriv: &keyspb.PrivateKey{Der: []byte("privkeybytes")},
+			},
+		},
+		{
+			desc:  "Duplicate DomainID",
+			write: true,
+			d: adminstorage.Domain{
+				Domain:  "testdomain",
+				MapID:   1,
+				LogID:   2,
+				VRF:     &keyspb.PublicKey{Der: []byte("pubkeybytes")},
+				VRFPriv: &keyspb.PrivateKey{Der: []byte("privkeybytes")},
+			},
+			wantWriteErr: true,
+		},
+		{
+			desc: "Delete",
+			d: adminstorage.Domain{
+				Domain:  "testdomain",
+				MapID:   1,
+				LogID:   2,
+				VRF:     &keyspb.PublicKey{Der: []byte("pubkeybytes")},
+				VRFPriv: &keyspb.PrivateKey{Der: []byte("privkeybytes")},
+			},
+			setDelete:   true,
+			isDeleted:   true,
+			readDeleted: false,
+			wantReadErr: true,
+		},
+		{
+			desc: "Read deleted",
+			d: adminstorage.Domain{
+				Domain:  "testdomain",
+				MapID:   1,
+				LogID:   2,
+				VRF:     &keyspb.PublicKey{Der: []byte("pubkeybytes")},
+				VRFPriv: &keyspb.PrivateKey{Der: []byte("privkeybytes")},
+			},
+			setDelete:   true,
+			isDeleted:   true,
+			readDeleted: true,
+			wantReadErr: false,
+		},
+		{
+			desc: "Undelete",
+			d: adminstorage.Domain{
+				Domain:  "testdomain",
+				MapID:   1,
+				LogID:   2,
+				VRF:     &keyspb.PublicKey{Der: []byte("pubkeybytes")},
+				VRFPriv: &keyspb.PrivateKey{Der: []byte("privkeybytes")},
+			},
+			setDelete:   true,
+			isDeleted:   false,
+			readDeleted: false,
+			wantReadErr: false,
+		},
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
```

Comment:
Any specific reason for using Run as opposed to running the test cases sequentially?

Reply:
`t.Run` is a nice way of notating each subtest with the description string without prepending each error message with the description.

Reply:
That's pretty cool, I didn't know about that :)